GVFS/GVFS.Common/GVFSEnlistmentHealthCalculator.cs

```diff
@@ -0,0 +1,240 @@
+        {
+            int gitTrackedFilesCount = 0;
+            int placeholderCount = 0;
+            int modifiedPathsCount = 0;
+            Dictionary<string, int> gitTrackedFilesDirectoryTally = new Dictionary<string, int>();
+            Dictionary<string, int> hydratedFilesDirectoryTally = new Dictionary<string, int>();
+
+            if (!parentDirectory.EndsWith(GVFSConstants.GitPathSeparatorString) && parentDirectory.Length > 0)
+            {
+                parentDirectory += GVFSConstants.GitPathSeparator;
+            }
+
+            if (parentDirectory.StartsWith(GVFSConstants.GitPathSeparatorString))
+            {
+                parentDirectory = parentDirectory.TrimStart(GVFSConstants.GitPathSeparator);
+            }
+
+            gitTrackedFilesCount += this.CategorizePaths(this.gitPaths, gitTrackedFilesDirectoryTally, parentDirectory, isFile: true);
+            placeholderCount += this.CategorizePaths(this.placeholderFolderPaths, hydratedFilesDirectoryTally, parentDirectory, isFile: false);
+            placeholderCount += this.CategorizePaths(this.placeholderFilePaths, hydratedFilesDirectoryTally, parentDirectory, isFile: true);
+            modifiedPathsCount += this.CategorizePaths(this.modifiedFolderPaths, hydratedFilesDirectoryTally, parentDirectory, isFile: false);
+            modifiedPathsCount += this.CategorizePaths(this.modifiedFilePaths, hydratedFilesDirectoryTally, parentDirectory, isFile: true);
+
+            Dictionary<string, decimal> mostHydratedDirectories = new Dictionary<string, decimal>();
+
+            // Map directory names to the corresponding health data from gitTrackedFilesDirectoryTally and hydratedFilesDirectoryTally
+            foreach (KeyValuePair<string, int> pair in gitTrackedFilesDirectoryTally)
+            {
+                if (hydratedFilesDirectoryTally.TryGetValue(pair.Key, out int hydratedFiles))
+                {
+                    // In-lining this for now until a better "health" calculation is created
+                    // Another possibility is the ability to pass a function to use for health (might not be applicable)
+                    mostHydratedDirectories.Add(pair.Key, (decimal)hydratedFiles / (decimal)pair.Value);
+                }
+                else
+                {
+                    mostHydratedDirectories.Add(pair.Key, 0);
+                }
+            }
+
+            // The list of directories which will be passed into the data object
+            List<ValueTuple<string, decimal>> outputDirectories = new List<ValueTuple<string, decimal>>();
+
+            // Create a list of the most hydrated directories in sorted order of specified length to return
+            foreach (KeyValuePair<string, decimal> pair in mostHydratedDirectories
+                .OrderByDescending(kp => kp.Value)
+                .ToList<KeyValuePair<string, decimal>>())

```

Comment:
Is this `ToList<...>(...)` necessary?

Reply:
I was concerned that since the data structure is a dictionary, the entries might be reordered after being sorted but upon farther exploration it looks like linq uses an intermediate data structure which is what the for loop will iterate on 👍